File: code/ch13/13.3.2.best_time_to_buy_and_sell_stock_with_fee.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        if (prices.size() <= 1)
            return 0;
        
        int pre_buy  = -prices[0];
        int pre_sell = 0;
        
        for (int i = 1; i < prices.size(); i++) {
            int buy  = max(pre_buy, pre_sell - prices[i]);
            int sell = max(pre_buy + prices[i] - fee, pre_sell);
            pre_buy  = buy;
            pre_sell = sell;
        }
        
        return pre_sell;
    }
};

File: code/ch13/13.3.2.best_time_to_buy_and_sell_stock_with_fee.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        if len(prices) <= 1:
            return 0
        pre_buy = -prices[0]
        pre_sell = 0
        for i in range(1, len(prices)):
            buy = max(pre_buy, pre_sell - prices[i])
            sell = max(pre_buy + prices[i] - fee, pre_sell)
            pre_buy = buy
            pre_sell = sell
        return pre_sell
